Tests/iaas/volume-backup/volume-backup-tester.py

```diff
@@ -64,133 +80,160 @@ def test_backup(conn: openstack.connection.Connection,
     )
     volume_id = conn.block_storage.find_volume(restored_volume_name).id
     while conn.block_storage.get_volume(volume_id).status != "available":
         time.sleep(1.0)
         seconds_waited += 1
-        assert seconds_waited < timeout, (
+        ensure(
+            seconds_waited < timeout,
             f"Timeout reached while waiting for restored volume reach status "
             f"'available' (volume id: {volume_id}) after {seconds_waited} "
             f"seconds"
         )
-    print("Restore volume from backup: PASS")
+    logging.info("Restore volume from backup: PASS")
 
 
 def cleanup(conn: openstack.connection.Connection, prefix=DEFAULT_PREFIX,
-            timeout=WAIT_TIMEOUT):
+            timeout=WAIT_TIMEOUT) -> bool:
     """
     Looks up volume and volume backup resources matching the given prefix and
     deletes them.
+    Returns False if there were any errors during cleanup which might leave
+    resources behind. Otherwise returns True to indicate cleanup success.
     """
 
     def wait_for_resource(resource_type: str, resource_id: str,
-                          expected_status="available") -> None:
+                          expected_status=("available", )) -> None:
         seconds_waited = 0
         get_func = getattr(conn.block_storage, f"get_{resource_type}")
-        while get_func(resource_id).status != expected_status:
+        while get_func(resource_id).status not in expected_status:
             time.sleep(1.0)
             seconds_waited += 1
-            assert seconds_waited < timeout, (
+            ensure(
+                seconds_waited < timeout,
                 f"Timeout reached while waiting for {resource_type} during "
-                f"cleanup to be in status '{expected_status}' "
+                f"cleanup to be in status {expected_status} "
                 f"({resource_type} id: {resource_id}) after {seconds_waited} "
                 f"seconds"
             )
 
-    print(f"\nPerforming cleanup for resources with the "
-          f"'{prefix}' prefix ...")
+    logging.info(f"Performing cleanup for resources with the "
+                 f"'{prefix}' prefix ...")
 
+    cleanup_was_successful = True
     backups = conn.block_storage.backups()
     for backup in backups:
         if backup.name.startswith(prefix):
             try:
-                wait_for_resource("backup", backup.id)
+                wait_for_resource(
+                    "backup", backup.id,
+                    expected_status=("available", "error")
+                )
             except openstack.exceptions.ResourceNotFound:
                 # if the resource has vanished on
                 # its own in the meantime ignore it
                 continue
-            print(f"↳ deleting volume backup '{backup.id}' ...")
-            conn.block_storage.delete_backup(backup.id)
+            except ConformanceTestException as e:
+                logging.warning(str(e))
```

Comment:
Maybe add a comment here why we need not set `cleanup_was_successful` to False here.
We don't issue the `delete_backup` command for this backup, so (in general, if no concurrent miracle happens) this backup won't go away, which will lead to a time out in the next loop (lines 239 and following).
On second thought: the exception that is caught here... does it really mean we shouldn't issue the deletion command? We could still try, couldn't we? (Otherwise it is a bit mean towards the user to wantonly provoke a timeout!)

Reply:
I added a comment.

> does it really mean we shouldn't issue the deletion command? We could still try, couldn't we?

In the wait command we expect either "available" or "error" state. The exception caught here only happens if it is none of those states. Those states are the only ones that allow deletion:

```
stack@devstack:~/devstack$ openstack volume backup delete 3b21f286-2db2-4e8a-9910-b3f2416d4317
Failed to delete backup with name or ID '3b21f286-2db2-4e8a-9910-b3f2416d4317':
BadRequestException: 400: Client Error for url: http://<redacted>/volume/
v3/47d9a75089464a7ea72506084e434fd3/backups/3b21f286-2db2-4e8a-9910-b3f2416d4317,
Invalid backup: Backup status must be available or error
1 of 1 backups failed to delete.
```

Reply:
> Those states are the only ones that allow deletion

Well, now I know. :)

Reply:
It could be a future improvement of the script to not wait for backups to go away that cannot even be deleted. But it's not critical for now.